p4-16/psa/PSA.mdk

```diff
@@ -3128,3 +3141,165 @@ the performance of these protocols.
+(mentioned in Section [#sec-action-selector]), a developer wishing to
+take advantage of this in a portable way may need to modify one or
+more of the actions used in their tables that use action selectors.
+
+For example, if in the LAG port selection example mentioned earlier,
+there was only one action for a table `lag` defined like this:
+
+```
+    action set_output_port (PortId_t p) {
+        user_meta.out_port = p;
+    }
+    ActionProfile(128) ap;
+    table lag {
+        key = {
+            // ... key fields go here ...
+        }
+        actions = { set_output_port; }
+        psa_implementation = ap;
+    }
+
+    control cIngress (
+                inout headers hdr,
+                inout metadata user_meta,
+                in    psa_ingress_input_metadata_t  istd_meta,
+                inout psa_ingress_output_metadata_t ostd_meta
+    ) {
+        apply {
+            // ... earlier ingress code goes here ...
+            lag.apply();
+            send_to_port(ostd_meta, user_meta.out_port);
+            // ... later ingress code goes here ...
+        }
+    }
+```
+
+with a single action parameter equal to a physical port number of the
+device, one of the following approaches could be used, and of course
+there are likely to be other approaches not mentioned here.
+
++ Approach 1: Use an invalid port number
+
+Pick a value of type `PortId_t` that corresponds to no valid physical
+port of the device (TBD whether PSA should define such a value with a
+name -- it currently doesn't guarantee that such a value even exists
+for type `PortId_t`) and use that value for the default action of an
+empty group.  In the code after `lag.apply`, add an `if` statement
+checking for that invalid value, like this:
+
+```
+        apply {
+            // ... earlier ingress code goes here ...
+            lag.apply();
+            if (user_meta.out_port == PORT_INVALID_VALUE) {
+                ingress_drop(ostd_meta);
+            } else {
+                send_to_port(ostd_meta, user_meta.out_port);
+            }
+            // ... later ingress code goes here ...
+        }
+```
+
++ Approach 2: Add extra parameters to the action
+
+In this case, add a 1-bit parameter indicating whether to drop the
+packet later.  An `if` statement is still needed after applying the
+table.
+
+```
+    action set_output_port (PortId_t p, bit<1> drop) {
+        user_meta.out_port = p;
+        user_meta.drop = drop;
+    }
+
+    // ...
+
+        apply {
+            // ... earlier ingress code goes here ...
+            lag.apply();
+            if (user_meta.drop == 1) {
+                ingress_drop(ostd_meta);
+            } else {
+                send_to_port(ostd_meta, user_meta.out_port);
+            }
+            // ... later ingress code goes here ...
+        }
+```
+
+In either case, an implementation might also support putting the `if`
+statement inside of the action `set_output_port`, but this is not
+required by PSA.
```

Comment:
did you want to mention "fast-failover" as one possible reason for the group to become empty without action from the switch or controller software? That's a valid way to implement the P4Runtime "watch" attribute if the HW supports it. In this case, the behavior may be undefined until the software is notified that the group is empty and can take action.

Reply:
Hashing this out in live conversation might be better, but I am also confused by this sentence of your comment: "In this case, the behavior may be undefined until the software is notified that the group is empty and can take action."  The reason for this new appendix is to _avoid_ undefined behavior, even if the data plane does not have predictable behavior for empty action selector groups.  The combined "data plane + P4Runtime agent" system can still have predictable portable behavior for times when the "action selector group is empty, as far as the controller knows".